llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp

```diff
@@ -0,0 +1,321 @@
+//===- SIConvertWaveSize.cpp - Automatically converts wave32 kernels to wave64
+//---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 WITH LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+// Small short living kernels may become waveslot limited.
+// To work around the problem an optimization is proposed to convert such
+// kernels from wave32 to wave64 automatically.These kernels shall conform to a
+// strict set of limitations and satisfy profitability conditions.
+//
+// 1. A kernel shall have no function calls as we cannot analyze call stack
+// requirements (nor will it fall into a category of short living kernels
+// anyway).
```

Comment:
If you did this in the attributor and could convert the entire call graph this wouldn't be an issue

Reply:
Not needed though. A cost of one call will make the whole transformation useless.

Reply:
The call itself isn't necessarily expensive, but the attributor would also take care of most of these queries even within a single kernel

Reply:
We are talking about quite low latency threshold.